docs/proposals/volume-provisioning.md

```diff
@@ -0,0 +1,500 @@
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+## Abstract
+
+Real Kubernetes clusters have a variety of volumes which differ widely in
+size, iops performance, retention policy, and other characteristics.
+Administrators need a way to dynamically provision volumes of these different
+types to automatically meet user demand.
+
+A new mechanism called 'storage classes' is proposed to provide this
+capability.  Storage classes build upon the label selection feature added in
+Kubernetes 1.3.
+
+## Motivation
+
+In Kubernetes 1.2, an alpha form of limited dynamic provisioning was added
+that allows a single volume type to be provisioned in clouds that offer
+special volume types.
+
+In Kubernetes 1.3, a label selector was added to persistent volume claims to
+allow administrators to create a taxonomy of volumes based on the
+characteristics important to them, and to allow users to make claims on those
+volumes based on those characteristics.  This allows flexibility when claiming
+existing volumes; the same flexibility is needed when dynamically provisioning
+volumes.
+
+After gaining experience with dynamic provisioning after the 1.2 release, we
+want to create a more flexible feature that allows configuration of how
+different storage classes are provisioned and supports provisioning multiple
+types of volumes within a single cloud.
+
+### Out-of-tree provisioners
+
+One of our goals is to enable administrators to create out-of-tree
+provisioners, that is, provisioners whose code does not live in the Kubernetes
+project.  Our experience since the 1.2 release with dynamic provisioning has
+shown that it is impossible to anticipate every aspect and manner of
+provisioning that administrators will want to perform.  The proposed design
+should not prevent future work to allow out-of-tree provisioners.
+
+## Design
+
+We propose that:
+
+1.  A new API group called `storage` be created to hold the a `StorageClass`
+    API resource
+
+2.  `PersistentVolume.Spec.Class` attribute is added to volumes. This attribute
+    is optional and specifies which `StorageClass` instance represents
+    storage characteristics of a particular PV.
+
+3.  `PersistentVolume` instances should be labelled with configuration
```

Comment:
For consistency. In extreme case, the API allows StorageClass to bey empty, just pointing to a provisioner plugin, allowing users to select SSD/HDD via `claim.Spec.Selector` in the same way as they can configure region. And if the controller should look for existing volumes first, all PVs must be labelled properly also with SSD/HDD labels.

Note that I am not advocating for this kind of use case, I am just making the API consistent with itself. For Kubernetes, storage properties are just opaque labels and it should treat volume type in the same way as region/zone.

Or do we really want to make some properties to be available only in StorageClass and not in claim.Selector?